File: applications/rtkfourdfdk/rtkfourdfdk.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkfourdfdk_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkConfiguration.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkDisplacedDetectorForOffsetFieldOfViewImageFilter.h"
#include "rtkParkerShortScanImageFilter.h"
#include "rtkFDKConeBeamReconstructionFilter.h"
#ifdef RTK_USE_CUDA
#  include "rtkCudaDisplacedDetectorImageFilter.h"
// TODO #  include "rtkCudaDisplacedDetectorForOffsetFieldOfViewImageFilter.h"
#  include "rtkCudaParkerShortScanImageFilter.h"
#  include "rtkCudaFDKConeBeamReconstructionFilter.h"
#endif
#include "rtkSelectOneProjectionPerCycleImageFilter.h"
#include <itkImageRegionConstIterator.h>
#include <itkStreamingImageFilter.h>
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkfourdfdk, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

  using CPUOutputImageType = itk::Image<OutputPixelType, Dimension>;
#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
#else
  using OutputImageType = CPUOutputImageType;
#endif

  // Projections reader
  using ReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto reader = ReaderType::New();
  rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtkfourdfdk>(reader, args_info);

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Part specific to 4D
  auto selector = rtk::SelectOneProjectionPerCycleImageFilter<OutputImageType>::New();
  selector->SetInput(reader->GetOutput());
  selector->SetInputGeometry(geometry);
  selector->SetSignalFilename(args_info.signal_arg);

  // Check on hardware parameter
#ifndef RTK_USE_CUDA
  if (!strcmp(args_info.hardware_arg, "cuda"))
  {
    std::cerr << "The program has not been compiled with cuda option" << std::endl;
    return EXIT_FAILURE;
  }
#endif

  // Displaced detector weighting
#ifdef RTK_USE_CUDA
  using DDFType = rtk::CudaDisplacedDetectorImageFilter;
#else
  using DDFType = rtk::DisplacedDetectorForOffsetFieldOfViewImageFilter<OutputImageType>;
#endif
  rtk::DisplacedDetectorImageFilter<OutputImageType>::Pointer ddf;
  if (!strcmp(args_info.hardware_arg, "cuda"))
    ddf = DDFType::New();
  else
    ddf = rtk::DisplacedDetectorForOffsetFieldOfViewImageFilter<OutputImageType>::New();
  ddf->SetInput(selector->GetOutput());
  ddf->SetGeometry(selector->GetOutputGeometry());

  // Short scan image filter
  using PSSFCPUType = rtk::ParkerShortScanImageFilter<OutputImageType>;
#ifdef RTK_USE_CUDA
  using PSSFType = rtk::CudaParkerShortScanImageFilter;
#else
  using PSSFType = rtk::ParkerShortScanImageFilter<OutputImageType>;
#endif
  PSSFCPUType::Pointer pssf;
  if (!strcmp(args_info.hardware_arg, "cuda"))
    pssf = PSSFType::New();
  else
    pssf = PSSFCPUType::New();
  pssf->SetInput(ddf->GetOutput());
  pssf->SetGeometry(selector->GetOutputGeometry());
  pssf->InPlaceOff();

  // Create one frame of the reconstructed image
  using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
  auto constantImageSource = ConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkfourdfdk>(constantImageSource, args_info);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(constantImageSource->Update())

  // This macro sets options for fdk filter which I can not see how to do better
  // because TFFTPrecision is not the same, e.g. for CPU and CUDA (SR)
#define SET_FELDKAMP_OPTIONS(f)                                   \
  f->SetInput(0, constantImageSource->GetOutput());               \
  f->SetInput(1, pssf->GetOutput());                              \
  f->SetGeometry(selector->GetOutputGeometry());                  \
  f->GetRampFilter()->SetTruncationCorrection(args_info.pad_arg); \
  f->GetRampFilter()->SetHannCutFrequency(args_info.hann_arg);    \
  f->GetRampFilter()->SetHannCutFrequencyY(args_info.hannY_arg);  \
  f->SetProjectionSubsetSize(args_info.subsetsize_arg)

  // FDK reconstruction filtering
  using FDKCPUType = rtk::FDKConeBeamReconstructionFilter<OutputImageType>;
  FDKCPUType::Pointer feldkamp;
#ifdef RTK_USE_CUDA
  using FDKCUDAType = rtk::CudaFDKConeBeamReconstructionFilter;
  FDKCUDAType::Pointer feldkampCUDA;
#endif
  itk::Image<OutputPixelType, Dimension> * pfeldkamp = nullptr;
  if (!strcmp(args_info.hardware_arg, "cpu"))
  {
    feldkamp = FDKCPUType::New();
    SET_FELDKAMP_OPTIONS(feldkamp);

    pfeldkamp = feldkamp->GetOutput();
  }
#ifdef RTK_USE_CUDA
  else if (!strcmp(args_info.hardware_arg, "cuda"))
  {
    feldkampCUDA = FDKCUDAType::New();
    SET_FELDKAMP_OPTIONS(feldkampCUDA);
    pfeldkamp = feldkampCUDA->GetOutput();
  }
#endif

  // Streaming depending on streaming capability of writer
  auto streamerBP = itk::StreamingImageFilter<CPUOutputImageType, CPUOutputImageType>::New();
  streamerBP->SetInput(pfeldkamp);
  streamerBP->SetNumberOfStreamDivisions(args_info.divisions_arg);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(streamerBP->UpdateOutputInformation())

  // Create empty 4D image
  using FourDOutputImageType = itk::Image<OutputPixelType, Dimension + 1>;
  using FourDConstantImageSourceType = rtk::ConstantImageSource<FourDOutputImageType>;
  auto fourDConstantImageSource = FourDConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<FourDConstantImageSourceType, args_info_rtkfourdfdk>(fourDConstantImageSource,
                                                                                          args_info);

  // GenGetOpt can't handle default arguments for multiple arguments like size or spacing.
  // The only default it accepts is to set all components of a multiple argument to the same value.
  // Default size is 256^4, ie the number of reconstructed instants is 256. It has to be set to a more reasonable
  // value which is why a "frames" argument is introduced
  FourDConstantImageSourceType::SizeType fourDInputSize(fourDConstantImageSource->GetSize());
  fourDInputSize[3] = args_info.frames_arg;
  fourDConstantImageSource->SetSize(fourDInputSize);

  TRY_AND_EXIT_ON_ITK_EXCEPTION(fourDConstantImageSource->Update())

  // Go over each frame, reconstruct 3D frame and paste with iterators in 4D image
  for (int f = 0; f < args_info.frames_arg; f++)
  {
    if (args_info.verbose_flag)
      std::cout << "Reconstructing frame #" << f << "..." << std::endl;
    selector->SetPhase(f / (double)args_info.frames_arg);
    TRY_AND_EXIT_ON_ITK_EXCEPTION(streamerBP->UpdateLargestPossibleRegion())

    FourDConstantImageSourceType::OutputImageRegionType region;
    region = fourDConstantImageSource->GetOutput()->GetLargestPossibleRegion();
    region.SetIndex(3, f);
    region.SetSize(3, 1);

    itk::ImageRegionIterator<FourDOutputImageType> it4D(fourDConstantImageSource->GetOutput(), region);
    itk::ImageRegionIterator<CPUOutputImageType>   it3D(streamerBP->GetOutput(),
                                                      streamerBP->GetOutput()->GetLargestPossibleRegion());
    while (!it3D.IsAtEnd())
    {
      it4D.Set(it3D.Get());
      ++it4D;
      ++it3D;
    }
  }

  // Write
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(fourDConstantImageSource->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkfourdfdk/rtkfourdfdk.py
#!/usr/bin/env python
import argparse
import sys
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Reconstructs a 3D + time (4D) sequence using FDK with one projection per respiratory cycle in each frame."
    )
    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--signal",
        help="File containing the phase of each projection",
        type=str,
        required=True,
    )
    parser.add_argument(
        "--hardware",
        help="Hardware used for computation",
        choices=["cpu", "cuda"],
        default="cpu",
    )
    parser.add_argument(
        "--divisions",
        "-d",
        help="Streaming option: number of stream divisions of the 3D reconstruction per frame",
        type=int,
        default=1,
    )
    parser.add_argument(
        "--subsetsize",
        help="Streaming option: number of projections processed at a time",
        type=int,
        default=16,
    )
    parser.add_argument(
        "--nodisplaced",
        help="Disable the displaced detector filter",
        action="store_true",
    )
    # Ramp filter
    ramp = parser.add_argument_group("Ramp filter")
    ramp.add_argument(
        "--pad",
        help="Data padding parameter to correct for truncation",
        type=float,
        default=0.0,
    )
    ramp.add_argument(
        "--hann",
        help="Cut frequency for Hann window in ]0,1] (0.0 disables it)",
        type=float,
        default=0.0,
    )
    ramp.add_argument(
        "--hannY",
        help="Cut frequency for Hann window in ]0,1] (0.0 disables it)",
        type=float,
        default=0.0,
    )

    # RTK common groups
    rtk.add_rtkinputprojections_group(parser)
    rtk.add_rtk4Doutputimage_group(parser)
    return parser


def process(args_info: argparse.Namespace):
    OutputPixelType = itk.F
    OutputImageType = itk.Image[OutputPixelType, 3]

    # Check on hardware parameter
    if args_info.hardware == "cuda" and not hasattr(itk, "CudaImage"):
        print("The program has not been compiled with cuda option")
        sys.exit(1)

    # Projections reader
    reader = rtk.ProjectionsReader[OutputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(reader, args_info)

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry information from {args_info.geometry}...")
    geometry = rtk.read_geometry(args_info.geometry)

    # Part specific to 4D
    selector = rtk.SelectOneProjectionPerCycleImageFilter[OutputImageType].New()
    selector.SetInput(reader.GetOutput())
    selector.SetInputGeometry(geometry)
    selector.SetSignalFilename(args_info.signal)

    # Displaced detector weighting
    if args_info.hardware == "cuda":
        ddf = rtk.CudaDisplacedDetectorImageFilter.New()
        ddf.SetInput(itk.cuda_image_from_image(selector.GetOutput()))
    else:
        ddf = rtk.DisplacedDetectorForOffsetFieldOfViewImageFilter[
            OutputImageType
        ].New()
        ddf.SetInput(selector.GetOutput())
    ddf.SetGeometry(selector.GetOutputGeometry())
    ddf.SetDisable(args_info.nodisplaced)

    # Short scan image filter
    if args_info.hardware == "cuda":
        pssf = rtk.CudaParkerShortScanImageFilter.New()
    else:
        pssf = rtk.ParkerShortScanImageFilter[OutputImageType].New()
    pssf.SetInput(ddf.GetOutput())
    pssf.SetGeometry(selector.GetOutputGeometry())
    pssf.InPlaceOff()

    # Create one frame of the reconstructed image
    constantImageSource = rtk.ConstantImageSource[OutputImageType].New()
    rtk.SetConstantImageSourceFromArgParse(constantImageSource, args_info)
    constantImageSource.Update()

    # FDK reconstruction filtering
    if args_info.hardware == "cuda":
        feldkamp = rtk.CudaFDKConeBeamReconstructionFilter.New()
        feldkamp.SetInput(0, itk.cuda_image_from_image(constantImageSource.GetOutput()))
        feldkamp.SetInput(1, pssf.GetOutput())
    else:
        feldkamp = rtk.FDKConeBeamReconstructionFilter[OutputImageType].New()
        feldkamp.SetInput(0, constantImageSource.GetOutput())
        feldkamp.SetInput(1, pssf.GetOutput())
    feldkamp.SetGeometry(selector.GetOutputGeometry())
    feldkamp.GetRampFilter().SetTruncationCorrection(args_info.pad)
    feldkamp.GetRampFilter().SetHannCutFrequency(args_info.hann)
    feldkamp.GetRampFilter().SetHannCutFrequencyY(args_info.hannY)
    feldkamp.SetProjectionSubsetSize(args_info.subsetsize)

    # Streaming depending on streaming capability of writer
    streamerBP = itk.StreamingImageFilter[OutputImageType, OutputImageType].New()
    streamerBP.SetInput(feldkamp.GetOutput())
    streamerBP.SetNumberOfStreamDivisions(args_info.divisions)
    # Prefer streaming splits along Z for cache friendliness (like rtkfdk)
    splitter = itk.ImageRegionSplitterDirection.New()
    splitter.SetDirection(2)
    streamerBP.SetRegionSplitter(splitter)
    # Initialize meta-information once
    streamerBP.UpdateOutputInformation()

    # Create empty 4D image
    fourDConstantImageSource = rtk.ConstantImageSource[
        itk.Image[OutputPixelType, 4]
    ].New()
    rtk.SetConstantImageSourceFromArgParse(fourDConstantImageSource, args_info)
    fourDInputSize = fourDConstantImageSource.GetSize()

    fourDInputSize[3] = args_info.frames
    fourDConstantImageSource.SetSize(fourDInputSize)
    fourDConstantImageSource.Update()

    # Go over each frame, reconstruct 3D frame and paste with NumPy views into the 4D image
    volOut = fourDConstantImageSource.GetOutput()
    arr4d = itk.GetArrayViewFromImage(volOut)
    num_frames = args_info.frames
    for f in range(num_frames):
        if args_info.verbose:
            print(f"Reconstructing frame #{f}...")
        selector.SetPhase(f / float(num_frames))
        streamerBP.UpdateLargestPossibleRegion()

        arr3d = itk.GetArrayViewFromImage(streamerBP.GetOutput())
        arr4d[f] = arr3d

    # Write
    if args_info.verbose:
        print(f"Writing output to {args_info.output}...")
    writer = itk.ImageFileWriter[itk.Image[OutputPixelType, 4]].New()
    writer.SetFileName(args_info.output)
    writer.SetInput(volOut)
    writer.SetNumberOfStreamDivisions(args_info.divisions)
    writer.Update()


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
